### PR TITLE
Text to SQL Bedrock agent

### PR DESCRIPTION
**Pull request by iut62elec**
*Issue #, if available:*

*Description of changes:*
This repository contains the necessary files to set up and test a Text to SQL conversion using the Bedrock Agent with AWS services.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: agents/text_2_sql_bedrock_agent/__pycache__/config.cpython-310.pyc
Comment (mttanke) [inline agents/text_2_sql_bedrock_agent/__pycache__/config.cpython-310.pyc:1]: please remove the pycache files

File: agents/text_2_sql_bedrock_agent/__pycache__/config.cpython-39.pyc
Comment (mttanke) [inline agents/text_2_sql_bedrock_agent/__pycache__/config.cpython-39.pyc:1]: please remove the pycache files

File: agents/text_2_sql_bedrock_agent/data/TheHistoryofBaseball.zip
Comment (mttanke) [inline agents/text_2_sql_bedrock_agent/data/TheHistoryofBaseball.zip:1]: can the data be downloaded from a public repo? is it available for commercial usage?

File: agents/text_2_sql_bedrock_agent/dependencies/__pycache__/config.cpython-39.pyc
Comment (mttanke) [inline agents/text_2_sql_bedrock_agent/dependencies/__pycache__/config.cpython-39.pyc:1]: please remove the pycache files

File: agents/text_2_sql_bedrock_agent/Readme.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+## Authors:
+Pedram Jahangiri @jpedram, Sawyer Hirt @sawyehir, Suyin Wang @suyinwa, Zeek Granston @zeekg 
+
+
+## Prerequisites
+
+Before you begin, ensure you have the following:
+- An AWS account with the following permissions:
+  - `IAMFullAccess`
```
Comment (mttanke) [inline agents/text_2_sql_bedrock_agent/Readme.md:13]: are those really necessary?

File: agents/text_2_sql_bedrock_agent/Readme.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        - AWS CLI installed and configured
+- For AWS SageMaker 
+    - Make sure your domain has above permission 
+    - use Data Science 3.0 kernel in SageMaker Studio
+
+## Installation
+
+Clone the repository to your local machine or AWS environment:
+
+git clone git@ssh.gitlab.aws.dev:jpedram/text_2_sql_bedrock_agent.git
```
Comment (mttanke) [inline agents/text_2_sql_bedrock_agent/Readme.md:28]: internal link, needs to be fixed

File: agents/text_2_sql_bedrock_agent/dependencies/build_infrastructure.py [hunk 3]
```diff
@@ -0,0 +1,467 @@
...
+# ### Create S3 bucket and upload API Schema
+# 
+# Agents require an API Schema stored on s3. Let's create an S3 bucket to store the file and upload the file to the newly created bucket
+
+
+
+# Create S3 bucket for Open API schema
+s3bucket = s3_client.create_bucket(
+    Bucket=bucket_name
+    #CreateBucketConfiguration={ 'LocationConstraint': region }
```
Comment (mttanke) [inline agents/text_2_sql_bedrock_agent/dependencies/build_infrastructure.py:13]: can you add a code to detect the region before and decide if the location constraint is required?

File: agents/text_2_sql_bedrock_agent/dependencies/build_infrastructure.py [hunk 4]
```diff
@@ -0,0 +1,467 @@
...
+            AssumeRolePolicyDocument=assume_role_policy_document_json
+        )
+
+        # Pause to make sure role is created
+        time.sleep(10)
+except:
+    glue_iam_role = iam_client.get_role(RoleName=glue_role_name)
+
+policy_arns = [
+    'arn:aws:iam::aws:policy/AWSGlueConsoleFullAccess',
```
Comment (mttanke) [inline agents/text_2_sql_bedrock_agent/dependencies/build_infrastructure.py:72]: can we have a more granular permission set?

File: agents/text_2_sql_bedrock_agent/dependencies/build_infrastructure.py [hunk 5]
```diff
@@ -0,0 +1,467 @@
...
+        RoleName=glue_role_name,
+        PolicyArn=policy_arn
+    )
+  
+
+    #crawler = glue.get_crawler(
+#     Name='TheHistoryofBaseball'
+# )
+# pprint.pprint(crawler)
+#s3_target = "text-2-sql-agent-us-east-1-456667773660/data/TheHistoryofBaseball/"
```
Comment (mttanke) [inline agents/text_2_sql_bedrock_agent/dependencies/build_infrastructure.py:87]: remove account id

File: agents/text_2_sql_bedrock_agent/dependencies/build_infrastructure.py [hunk 6]
```diff
@@ -0,0 +1,467 @@
...
+    )
+
+    # Pause to make sure role is created
+    time.sleep(10)
+except:
+    lambda_iam_role = iam_client.get_role(RoleName=lambda_role_name)
+
+    
+policy_arns = [
+    'arn:aws:iam::aws:policy/AmazonAthenaFullAccess',
```
Comment (mttanke) [inline agents/text_2_sql_bedrock_agent/dependencies/build_infrastructure.py:192]: can the policies be less granular?

File: agents/text_2_sql_bedrock_agent/dependencies/build_infrastructure.py [hunk 7]
```diff
@@ -0,0 +1,467 @@
...
+
+bedrock_agent_bedrock_allow_policy_statement = {
+    "Version": "2012-10-17",
+    "Statement": [
+        {
+            "Sid": "AmazonBedrockAgentBedrockFoundationModelPolicy",
+            "Effect": "Allow",
+            "Action": "bedrock:InvokeModel",
+            "Resource": [
+                f"arn:aws:bedrock:{region}::foundation-model/anthropic.claude-v2:1"
```
Comment (mttanke) [inline agents/text_2_sql_bedrock_agent/dependencies/build_infrastructure.py:247]: can we parameterize this?

File: agents/text_2_sql_bedrock_agent/dependencies/build_infrastructure.py [hunk 8]
```diff
@@ -0,0 +1,467 @@
...
+of the table then create a sql query to answer the users question.
+Here is an example to query the table <example>SELECT * FROM thehistoryofbaseball.players LIMIT 10;</example> Do not use 
+quotes for the table name. Youir final answer should be in plain english."""
+
+response = bedrock_agent_client.create_agent(
+    agentName=agent_name,
+    agentResourceRoleArn=agent_role['Role']['Arn'],
+    description="Agent for performing sql queries.",
+    idleSessionTTLInSeconds=3600,
+    foundationModel="anthropic.claude-v2:1",
```
Comment (mttanke) [inline agents/text_2_sql_bedrock_agent/dependencies/build_infrastructure.py:343]: can you parametrize this?

File: agents/text_2_sql_bedrock_agent/dependencies/clean.py [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+    for obj in objects['Contents']:
+        s3_client.delete_object(Bucket=bucket_name, Key=obj['Key']) 
+s3_client.delete_bucket(Bucket=bucket_name)
+
+
+
+
+
+policy_arns = [
+    'arn:aws:iam::aws:policy/AmazonAthenaFullAccess',
```
Comment (mttanke) [inline agents/text_2_sql_bedrock_agent/dependencies/clean.py:156]: can those be less granular?

File: agents/text_2_sql_bedrock_agent/dependencies/clean.py [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+    'arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole'
+]  
+for policy_arn in policy_arns:
+    iam_client.detach_role_policy(
+        RoleName=lambda_role_name,
+        PolicyArn=policy_arn
+    )    
+
+policy_arns = [
+    'arn:aws:iam::aws:policy/AWSGlueConsoleFullAccess',
```
Comment (mttanke) [inline agents/text_2_sql_bedrock_agent/dependencies/clean.py:169]: can those be less granular?

File: agents/text_2_sql_bedrock_agent/dependencies/config.py [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+s3_loc = "s3://" + bucket_name + "/" + bucket_key
+s3_bucket=bucket_name
+db_loc = "s3://" + s3_bucket + "/db/"
+athena_result_loc = "s3://" + s3_bucket + "/athena_result/" 
+print(db_loc)
+#glue_crawler_name='TheHistoryOfBaseball'
+
+zip_data = "./data/TheHistoryofBaseball.zip"
+ext_data = "./data/extracted/"
+#s3_bucket = "text-2-sql-agent-us-east-1-456667773660"
```
Comment (mttanke) [inline agents/text_2_sql_bedrock_agent/dependencies/config.py:70]: this needs to be parametrize. We cannot commit account ids

**Review by mttanke — CHANGES_REQUESTED** [review]

File: agents/text_2_sql_bedrock_agent/dependencies/build_infrastructure.py (hunk 7 above)
Comment (iut62elec) [reply agents/text_2_sql_bedrock_agent/dependencies/build_infrastructure.py:247]: done

File: agents/text_2_sql_bedrock_agent/dependencies/build_infrastructure.py (hunk 8 above)
Comment (iut62elec) [reply agents/text_2_sql_bedrock_agent/dependencies/build_infrastructure.py:343]: done

File: agents/text_2_sql_bedrock_agent/dependencies/config.py (hunk 11 above)
Comment (iut62elec) [reply agents/text_2_sql_bedrock_agent/dependencies/config.py:70]: done

File: agents/text_2_sql_bedrock_agent/data/TheHistoryofBaseball.zip
Comment (iut62elec) [reply agents/text_2_sql_bedrock_agent/data/TheHistoryofBaseball.zip:1]: Data is from "kaggle" and it is under "[CC BY-SA 3.0](https://creativecommons.org/licenses/by-sa/3.0/)" which means "You are free to:
Share — copy and redistribute the material in any medium or format for any purpose, even commercially.
"
https://www.kaggle.com/datasets/seanlahman/the-history-of-baseball?resource=download

I don't think it is good idea to download it as we run because needs kaggle credentials. Also it is just 800kb

File: agents/text_2_sql_bedrock_agent/__pycache__/config.cpython-310.pyc
Comment (iut62elec) [reply agents/text_2_sql_bedrock_agent/__pycache__/config.cpython-310.pyc:1]: done

File: agents/text_2_sql_bedrock_agent/__pycache__/config.cpython-39.pyc
Comment (iut62elec) [reply agents/text_2_sql_bedrock_agent/__pycache__/config.cpython-39.pyc:1]: done

File: agents/text_2_sql_bedrock_agent/dependencies/__pycache__/config.cpython-39.pyc
Comment (iut62elec) [reply agents/text_2_sql_bedrock_agent/dependencies/__pycache__/config.cpython-39.pyc:1]: done

File: agents/text_2_sql_bedrock_agent/Readme.md (hunk 1 above)
Comment (iut62elec) [reply agents/text_2_sql_bedrock_agent/Readme.md]: corrected

File: agents/text_2_sql_bedrock_agent/dependencies/clean.py (hunk 9 above)
Comment (iut62elec) [reply agents/text_2_sql_bedrock_agent/dependencies/clean.py:159]: I will check, everything else has been addressed, except these two Athena and Glue full access
